crates/bin/pd/src/upgrade.rs

```diff
@@ -14,74 +14,95 @@ use crate::testnet::generate::TestnetConfig;
+                /* ---------- generate genesis ------------  */
+                let validators = migrated_state.validator_list().await?;
+                let app_state = genesis::Content {
+                    chain_content: ChainContent { chain_params },
+                    stake_content: StakeContent {
+                        validators: validators.into_iter().map(Into::into).collect(),
+                        ..Default::default()
+                    },
                     ..Default::default()
-                },
-                ..Default::default()
-            };
-            let mut genesis =
-                TestnetConfig::make_genesis(app_state.clone()).expect("can make genesis");
-            genesis.app_hash = app_hash
-                .0
-                .to_vec()
-                .try_into()
-                .expect("infaillible conversion");
-            genesis.initial_height = post_ugprade_height as i64;
-            genesis.genesis_time = tendermint::time::Time::now();
-            let checkpoint = [32u8; 32].to_vec();
-            let genesis = TestnetConfig::make_checkpoint(genesis, Some(checkpoint));
+                };
+                let mut genesis =
+                    TestnetConfig::make_genesis(app_state.clone()).expect("can make genesis");
+                genesis.app_hash = app_hash
+                    .0
+                    .to_vec()
+                    .try_into()
+                    .expect("infaillible conversion");
+                genesis.initial_height = post_ugprade_height as i64;
+                genesis.genesis_time = genesis_start.unwrap_or_else(|| {
+                    let now = tendermint::time::Time::now();
+                    tracing::info!(%now, "no genesis time provided, detecting a testing setup");
+                    now
+                });
+                let checkpoint = [32u8; 32].to_vec();
+                let genesis = TestnetConfig::make_checkpoint(genesis, Some(checkpoint));
 
-            let genesis_json = serde_json::to_string(&genesis).expect("can serialize genesis");
-            tracing::info!("genesis: {}", genesis_json);
-            let mut genesis_path = path_to_export.clone();
-            genesis_path.push("genesis.json");
-            std::fs::write(genesis_path, genesis_json).expect("can write genesis");
+                let genesis_json = serde_json::to_string(&genesis).expect("can serialize genesis");
+                tracing::info!("genesis: {}", genesis_json);
+                let mut genesis_path = path_to_export.clone();
+                genesis_path.push("genesis.json");
+                std::fs::write(genesis_path, genesis_json).expect("can write genesis");
+
+                let mut validator_state_path = path_to_export.clone();
+                validator_state_path.push("priv_validator_state.json");
+                let fresh_validator_state =
+                    crate::testnet::generate::TestnetValidator::initial_state();
```

Comment:
Beautiful, this'll solve our problems! As an aside, is there any kind of chain upgrade where we _wouldn't_ want the validator state reset? I can't think of one.